scripts/run_e2e_tests.py

```diff
@@ -0,0 +1,473 @@
+    screeenshots_dir = os.path.join(os.pardir, 'protractor-screenshots')
+    if not os.path.isdir(screeenshots_dir):
+        return
+    python_utils.PRINT("""
+Note: If ADD_SCREENSHOT_REPORTER is set to true in
+core/tests/protractor.conf.js, you can view screenshots
+of the failed tests in ../protractor-screenshots/
+""")
+    os.rmdir(screeenshots_dir)
+
+
+def cleanup():
+    """Kill the running subprocesses and server fired in this program."""
+    dev_appserver_path = '%s/dev_appserver.py' % common.GOOGLE_APP_ENGINE_HOME
+    webdriver_download_path = '%s/downloads' % WEBDRIVER_HOME_PATH
+    if common.is_windows_os():
+        # In windows system, the java command line will use absolute path.
+        webdriver_download_path = os.path.abspath(webdriver_download_path)
+    processes_to_kill = [
+        '.*%s.*' % re.escape(dev_appserver_path),
+        '.*%s.*' % re.escape(webdriver_download_path)
+    ]
+    for p in SUBPROCESSES:
+        p.kill()
+
+    for p in processes_to_kill:
+        common.kill_processes_based_on_regex(p)
+
+
+def check_running_instance(*ports):
+    """Check if exists running oppia instance.
+
+    Args:
+        ports: list[int]. A list of ports that oppia instances may be running.
+
+    Return:
+        boolean represents whether the ports are open.
+    """
+    running = False
+    for port in ports:
+        if common.is_port_open(port):
+            python_utils.PRINT("""
+There is already a server running on localhost:%s.
+Please terminate it before running the end-to-end tests.
+    Exiting.
+            """ % port)
+            running = True
+            break
+    return running
+
+
+def wait_for_port(port):
+    """Wait until the port is open.
+
+    Args:
+        port: int. The port number to wait.
+    """
+    current_time = 0
+    while not common.is_port_open(port) and current_time < WAIT_PORT_TIMEOUT:
+        time.sleep(1)
+        current_time += 1
+    if current_time == WAIT_PORT_TIMEOUT and not common.is_port_open(port):
+        python_utils.PRINT(
+            'Failed to start server on port %s, exiting ...' % port)
+        sys.exit(1)
+
+
+def tweak_constant_file(constant_file, dev_mode):
+    """Change constant file based on the running mode. Only the `DEV_MODE` line
+    should be changed.
+
+    Args:
+        constant_file: str. File path to the constant file.
+        dev_mode: boolean. Represents whether the program is running on dev
+            mode.
+    """
+    pattern = '"DEV_MODE": .*'
+    replace = '"DEV_MODE": %s' % (
+        'true' if dev_mode else 'false')
+    common.inplace_replace_file(constant_file, pattern, replace)
+
+
+def tweak_feconf_file(feconf_file_path, enable_community_dashboard):
+    """Change feconf.py file based on whether the community dashboard is
+    enabled.
+
+    Args:
+        feconf_file_path: str. Path to the feconf.py file.
+        enable_community_dashboard: boolean. Represents whether community
```

Comment:
Here and elsewhere, follow existing docstring conventions. Use `bool`, `list(str)`, and make sure you actually have docstrings for all args. (See functions below as well; please proofread the file in its entirety.)

Reply:
Done